File: src/test/java/org/cloudfoundry/credhub/controller/v1/ManagementControllerTest.java
package org.cloudfoundry.credhub.controller.v1;

import org.cloudfoundry.credhub.CredentialManagerApp;
import org.cloudfoundry.credhub.util.AuthConstants;
import org.cloudfoundry.credhub.util.DatabaseProfileResolver;
import org.cloudfoundry.credhub.variables.ManagementVariables;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.context.TestPropertySource;
import org.springframework.test.context.junit4.SpringRunner;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.request.MockHttpServletRequestBuilder;
import org.springframework.test.web.servlet.setup.MockMvcBuilders;
import org.springframework.web.context.WebApplicationContext;

import static org.hamcrest.Matchers.is;
import static org.junit.Assert.assertThat;
import static org.springframework.http.MediaType.APPLICATION_JSON;
import static org.springframework.security.test.web.servlet.setup.SecurityMockMvcConfigurers.springSecurity;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.content;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@RunWith(SpringRunner.class)
@ActiveProfiles(value = "unit-test", resolver = DatabaseProfileResolver.class)
@SpringBootTest(classes = CredentialManagerApp.class)
@TestPropertySource(properties = "security.authorization.acls.enabled=false")
public class ManagementControllerTest {

  @Autowired
  private WebApplicationContext webApplicationContext;

  private MockMvc mockMvc;

  @Before
  public void setUp() {
    mockMvc = MockMvcBuilders
        .webAppContextSetup(webApplicationContext)
        .apply(springSecurity())
        .build();
  }

  @Test
  public void settingReadOnlyMode_updatesTheGlobalManagementVariable() throws Exception{
    MockHttpServletRequestBuilder request = post("/management")
        .header("Authorization", "Bearer " + AuthConstants.ALL_PERMISSIONS_TOKEN)
        .header("content-type", APPLICATION_JSON)
        .content("{\"read_only_mode\":\"true\"}");
    mockMvc.perform(request).andExpect(status().isOk());

    assertThat(ManagementVariables.readOnlyMode, is(true));

    request = post("/management")
        .header("Authorization", "Bearer " + AuthConstants.ALL_PERMISSIONS_TOKEN)
        .header("content-type", APPLICATION_JSON)
        .content("{\"read_only_mode\":\"false\"}");
    mockMvc.perform(request).andExpect(status().isOk());

    assertThat(ManagementVariables.readOnlyMode, is(false));
  }

  @Test
  public void gettingTheReadOnlyMode_returnsTheGlobalManagementVariable() throws Exception{
    ManagementVariables.readOnlyMode = true;

    MockHttpServletRequestBuilder request = get("/management")
        .header("Authorization", "Bearer " + AuthConstants.ALL_PERMISSIONS_TOKEN)
        .accept(APPLICATION_JSON);
    mockMvc.perform(request)
        .andExpect(status().isOk())
        .andExpect(content().contentTypeCompatibleWith(APPLICATION_JSON))
        .andExpect(jsonPath("$.read_only_mode").value(true));

    ManagementVariables.readOnlyMode = false;

    request = get("/management")
        .header("Authorization", "Bearer " + AuthConstants.ALL_PERMISSIONS_TOKEN)
        .accept(APPLICATION_JSON);
    mockMvc.perform(request)
        .andExpect(status().isOk())
        .andExpect(content().contentTypeCompatibleWith(APPLICATION_JSON))
        .andExpect(jsonPath("$.read_only_mode").value(false));
  }

  @Test
  public void providingAnInvalidRequestBody_returns400() throws Exception{
    MockHttpServletRequestBuilder request = post("/management")
        .header("Authorization", "Bearer " + AuthConstants.ALL_PERMISSIONS_TOKEN)
        .header("content-type", APPLICATION_JSON)
        .content("{\"read_only_mode\":\"pizza\"}");
    mockMvc.perform(request).andExpect(status().isBadRequest());
  }
}

File: backends/credhub/src/test/kotlin/org/cloudfoundry/credhub/controllers/autodocs/v1/management/ManagementControllerTest.kt
package org.cloudfoundry.credhub.controllers.autodocs.v1.management

import org.assertj.core.api.Assertions.assertThat
import org.cloudfoundry.credhub.management.ManagementController
import org.cloudfoundry.credhub.testhelpers.CredHubRestDocs
import org.cloudfoundry.credhub.testhelpers.MockMvcFactory
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.skyscreamer.jsonassert.JSONAssert
import org.springframework.http.MediaType
import org.springframework.restdocs.JUnitRestDocumentation
import org.springframework.restdocs.mockmvc.MockMvcRestDocumentation
import org.springframework.restdocs.mockmvc.RestDocumentationRequestBuilders.get
import org.springframework.restdocs.mockmvc.RestDocumentationRequestBuilders.post
import org.springframework.test.context.junit4.SpringRunner
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.result.MockMvcResultMatchers

@RunWith(SpringRunner::class)
class ManagementControllerTest {

    @Rule
    @JvmField
    val restDocumentation = JUnitRestDocumentation()

    lateinit var mockMvc: MockMvc
    lateinit var spyManagementService: SpyManagementService

    @Before
    fun setUp() {

        spyManagementService = SpyManagementService()

        val managementController = ManagementController(spyManagementService)
        mockMvc = MockMvcFactory.newSpringRestDocMockMvc(managementController, restDocumentation)
    }

    @Test
    fun GET__management_mode__returns_result() {
        spyManagementService.isReadOnlyMode_returns = false

        val mvcResult = mockMvc.perform(
            get(ManagementController.ENDPOINT)
                .contentType(MediaType.APPLICATION_JSON)
                .header("Authorization", "Bearer [some-token]")
        )
            .andExpect(MockMvcResultMatchers.status().isOk)
            .andExpect(MockMvcResultMatchers.content().contentTypeCompatibleWith(MediaType.APPLICATION_JSON))
            .andDo(
                MockMvcRestDocumentation.document(
                    CredHubRestDocs.DOCUMENT_IDENTIFIER
                )
            ).andReturn()

        val expectedResponseBody = """
            {
                "read_only_mode": false
            }
        """.trimIndent()
        val actualResponseBody = mvcResult.response.contentAsString
        JSONAssert.assertEquals(expectedResponseBody, actualResponseBody, true)
    }

    @Test
    fun POST__management_mode__returns_result() {
        val mvcResult = mockMvc.perform(
            post(ManagementController.ENDPOINT)
                .contentType(MediaType.APPLICATION_JSON)
                .header("Authorization", "Bearer [some-token]")
                .content(
                    """
                        {
                            "read_only_mode": true
                        }
                    """.trimIndent()
                )
        )
            .andExpect(MockMvcResultMatchers.status().isOk)
            .andExpect(MockMvcResultMatchers.content().contentTypeCompatibleWith(MediaType.APPLICATION_JSON))
            .andDo(
                MockMvcRestDocumentation.document(
                    CredHubRestDocs.DOCUMENT_IDENTIFIER
                )
            ).andReturn()

        assertThat(spyManagementService.toggleReadOnlyMode_calledWithShouldUseReadOnlyMode).isEqualTo(true)

        val expectedResponseBody = """
            {
                "read_only_mode": true
            }
        """.trimIndent()
        val actualResponseBody = mvcResult.response.contentAsString
        JSONAssert.assertEquals(expectedResponseBody, actualResponseBody, true)
    }
}
